hcl2template/types.build.go

```diff
@@ -254,16 +254,19 @@ func (p *Parser) decodeBuildConfig(block *hcl.Block, cfg *PackerConfig) (*BuildB
 			cfg.bucket.RegisterBuildForComponent(source.String())
 		}
 
-		// Known HCP Packer Image Datasource, whose id is the SourceImageId for some build.
-		const hcpDatasourceType string = "hcp-packer-image"
+		// Known HCP Packer Image Datasource, whose id is the SourceImageId and ChannelID for some build.
 		values := ectx.Variables[dataAccessor].AsValueMap()
 
-		dsValues, ok := values[hcpDatasourceType]
+		const hcpImageDatasourceType string = "hcp-packer-image"
+		imageDatasourceValues, ok := values[hcpImageDatasourceType]
 		if !ok {
 			return build, diags
 		}
 
-		for _, value := range dsValues.AsValueMap() {
+		const hcpIterationDatasourceType string = "hcp-packer-iteration"
+		iterationDatasourceValues, usingIterationDatasource := values[hcpIterationDatasourceType]
```

Comment:
This was a little confusing at first given the name of `usingIterationDatatsource`. I see now that this is your ok variable which you are using further down in the code. As opposed to having the nested for loop within the if block.  

What do you think about collecting all the values needed from this datasource if available and then use them further down if ChannelID is empty. 

```go
                interationToChannels := make(map[string]string)
		if iterationDatasourceValues, ok := values[hcpIterationDatasourceType]; ok {
		  for _, itValue := range iterationDatasourceValues.AsValueMap() {
					if !itValue.IsKnown() {
						continue
					}

					itValues := itValue.AsValueMap()
					k,v := itValues["id"].AsString(), itValues["channel_id"].AsString()
					interationToChannels[k] = v
		  }
	      }

             //.... 
             for _, value := range imageDatasourceValues.AsValueMap() {
             //....
               if sourceIteration.ChannelID  == "" {
                 sourceIteration.ChannelID = iterationToChannels[itID.AsString()]
               } 
           }
            
```

Reply:
I like your idea!! Done!